keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,790 @@
+for more examples.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Open Questions on Deployment Controller
+
+The Deployment API is open for discussion.  We put the field in Deployment/ReplicaSet because it is related to RolloutStrategy.
+It is not clear if `recreate` and/or `rollingupdate` need this API for both rollout options.
+
+Another open question is if we want to include Deployments in the initial release of this feature.  There is some discussion about releasing the Job API first and then follow up with Deployment.  
+
+We decided to define the APIs in this KEP as they can utilize the same implementation.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API Name Choices
+
+- TerminatingAsActive
+- ActiveUntilTerminal
+- DelayPodRecreationUntilTerminal
+- ?
+
+### Deployment/ReplicaSet API
+
+```golang
+// DeploymentSpec stores information about the strategy and rolling-update
+// behavior of a deployment.
+type DeploymentSpec struct {
+  ... 
+  // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+So we can count terminating pods separately from active or failed we need to include a new field in the ReplicaSetStatus and DeploymentStatus.
+
+```golang
+type DeploymentStatus struct {
+  ...
+  // Terminating replicas states the number of replicas that are terminating
+  // +optional
+  TerminatingReplicas *int32 
+}
+```
+
+```golang
+type ReplicaSetStatus struct {
+  ...
+  // Terminating replicas states the number of replicas that are terminating
+  // +optional
+  TerminatingReplicas *int32 
+}
+```
+
+In [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+there was a request to add this as part of the `DeploymentStrategy` field.  Generally, handling terminating pods as active can be useful in both RollingUpdates and Recreating rollouts.  Having this field for both strategies allows for handling of terminating pods in both cases.  
+
+Deployments create ReplicaSets so there is a need to add a field in the ReplicaSet as well.  Since ReplicaSets are not typically
+set by users, we should add a field to the ReplicaSet that is set from the DeploymentSpec.  
+
+```golang
+// ReplicaSetSpec is the specification of a ReplicaSet.
+// As the internal representation of a ReplicaSet, it must have
+// a Template set.
+type ReplicaSetSpec struct {
+  ...
+ // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+### Implementation
```

Comment:
The implementation part will be done in a deployment and not in a replica set and is probably not feasible to write it fully here. So IMO the expected behaviours and replicasets transitiions should be discussed mainly here. And then we can decide if some important parts of the implementation should be highlighted.